rllib/offline/feature_importance.py

```diff
@@ -1,29 +1,129 @@
+import copy
+import numpy as np
+import pandas as pd
 from typing import Callable, Dict, Any
+
+import ray
+from ray.data import Dataset
+
 from ray.rllib.policy import Policy
-from ray.rllib.policy.sample_batch import convert_ma_batch_to_sample_batch
-from ray.rllib.utils.annotations import override, DeveloperAPI
+from ray.rllib.policy.sample_batch import SampleBatch, convert_ma_batch_to_sample_batch
+from ray.rllib.utils.annotations import override, DeveloperAPI, ExperimentalAPI
 from ray.rllib.utils.typing import SampleBatchType
 from ray.rllib.offline.offline_evaluator import OfflineEvaluator
 
-import numpy as np
-import copy
-
 
 @DeveloperAPI
-def perturb_fn(batch: np.ndarray, index: int):
+def _perturb_fn(batch: np.ndarray, index: int):
     # shuffle the indexth column features
     random_inds = np.random.permutation(batch.shape[0])
     batch[:, index] = batch[random_inds, index]
 
 
+@ExperimentalAPI
+def _perturb_df(batch: pd.DataFrame, index: int):
+    obs_batch = np.vstack(batch["obs"].values)
+    _perturb_fn(obs_batch, index)
+    batch["perturbed_obs"] = list(obs_batch)
+    return batch
+
+
+def _compute_actions(
+    batch: pd.DataFrame,
+    policy_state: Dict[str, Any],
+    input_key: str = "",
+    output_key: str = "",
+):
+    """A custom local function to do batch prediction of a policy.
+
+    Given the policy state the action predictions are computed as a function of
+    `input_key` and stored in the `output_key` column.
+
+    Args:
+        batch: A sub-batch from the dataset.
+        policy_state: The state of the policy to use for the prediction.
+        input_key: The key to use for the input to the policy. If not given, the
+            default is SampleBatch.OBS.
+        output_key: The key to use for the output of the policy. If not given, the
+            default is "predicted_actions".
+
+    Returns:
+        The modified batch with the predicted actions added as a column.
+    """
+    if not input_key:
+        input_key = SampleBatch.OBS
+
+    policy = Policy.from_state(policy_state)
+    sample_batch = SampleBatch(
+        {
+            SampleBatch.OBS: np.vstack(batch[input_key].values),
+        }
+    )
+    actions, _, _ = policy.compute_actions_from_input_dict(sample_batch, explore=False)
+
+    if not output_key:
+        output_key = "predicted_actions"
+    batch[output_key] = actions
+
+    return batch
+
+
+@ray.remote
+def get_feature_importance_on_index(
+    dataset: ray.data.Dataset,
+    *,
+    index: int,
+    perturb_fn: Callable[[pd.DataFrame, int], None],
+    batch_size: int,
+    policy_state: Dict[str, Any],
+):
+    """A remote function to compute the feature importance of a given index.
+
+    Args:
+        dataset: The dataset to use for the computation. The dataset should have `obs`
+            and `actions` columns. Each record should be flat d-dimensional array.
```

Comment:
mention that we only perturb / study features in the last dimension?

Reply:
we compute it on all dims. I don't understand this comment.

Reply:
line 253 ``n_features = ds.take(1)[0][SampleBatch.OBS].shape[-1]``?

Reply:
oh that's just getting the number of features.

Reply:
but that's how many features we study right? when we perturb, we only shuffle along the last dim?

Reply:
all of them. for each feature dim, you need to shuffle that entire column in the dataset. and do batch inference.